helix-gateway/src/main/java/org/apache/helix/gateway/channel/HelixGatewayServicePollModeChannel.java

```diff
@@ -20,38 +20,195 @@
+  // It is the file pass if _participantConnectionStatusChannelType is FILE, grpc endpoint if it is GRPC_CLIENT
+  final Map<String, Map<String, String>> _livenessCheckEndpointMap;
+
+  ScheduledExecutorService _scheduler;
+
+  public HelixGatewayServicePollModeChannel(GatewayServiceManager manager, GatewayServiceChannelConfig config) {
+    _manager = manager;
+    _config = config;
+    _scheduler = Executors.newSingleThreadScheduledExecutor();
+    _participantConnectionStatusChannelType = _config.getParticipantConnectionChannelType();
+    _shardStateChannelType = _config.getShardStateChannelType();
+    _livenessCheckEndpointMap = _config.getParticipantLivenessEndpointMap();
+    _userCurrentStateFilePath = _config.getPollModeConfig(PARTICIPANT_CURRENT_STATE_PATH);
+    _targetStateFilePath = _config.getPollModeConfig(SHARD_TARGET_STATE_PATH);
+    _livenessResults = new HashMap<>();
+  }
 
-  public HelixGatewayServicePollModeChannel(GatewayServiceChannelConfig config) {
+  void fetchUpdates() {
+    // 1.  get the shard state change
+    Map<String, Map<String, Map<String, Map<String, String>>>> currentShardStates =
+        getChangedParticipantsCurrentState(_userCurrentStateFilePath);
+
+    Map<String, Map<String, Map<String, Map<String, String>>>> currentStateDiff = new HashMap<>();
+    for (String clusterName : currentShardStates.keySet()) {
+      Map<String, Map<String, Map<String, String>>> clusterDiffMap =
+          _manager.updateCacheWithNewCurrentStateAndGetDiff(clusterName, currentShardStates.get(clusterName));
+      if (clusterDiffMap == null || clusterDiffMap.isEmpty()) {
+        continue;
+      }
+      for (String instanceName : clusterDiffMap.keySet()) {
+        // if the instance is previously connected, send state change event
+        if (_livenessResults.get(clusterName) != null && _livenessResults.get(clusterName).get(instanceName)) {
+          logger.info("Host {} has state change, sending event to gateway manager", instanceName);
+          pushClientEventToGatewayManager(_manager,
+              StateTransitionMessageTranslateUtil.translateCurrentStateChangeToEvent(clusterName, instanceName,
+                  clusterDiffMap.get(instanceName)));
+        }
+      }
+      currentStateDiff.put(clusterName, clusterDiffMap);
+    }
+
+    // 2. fetch host health
+    for (String clusterName : _livenessCheckEndpointMap.keySet()) {
+      for (String instanceName : _livenessCheckEndpointMap.get(clusterName).keySet()) {
+        boolean prevLiveness =
+            _livenessResults.get(clusterName) != null && _livenessResults.get(clusterName).get(instanceName);
+        boolean liveness = fetchInstanceLivenessStatus(clusterName, instanceName);
+
+        if (prevLiveness && !liveness) {  // previously connected, now disconnected
+          logger.warn("Host {} is not healthy, sending event to gateway manager", instanceName);
+          pushClientEventToGatewayManager(_manager,
```

Comment:
Was the manager side queue guarantee the order? I am a little bit worry that the order can mess up the state.

Because here the prevlive and liveness are all client side reported result. Could there be an inconsistent the HelixManager connection state not match this?

Reply:
The connect state change queue is a PerKeyBlockingExecutor. Meaning all connect/disconnect events for one instance are executed in the order of receiving.

Reply:
It is not related to this PR but I am a little bit worry about the Executor. I would suggest to spin thread by your self instead of using single threaded thread pool. Because you are not dedup queue, any flip/flop issue can buster the pending queue of the pool

Reply:
How about we have one thread for each instance? So if one instance is keep connecting/disconnecting, wont impact others.

Reply:
It is blocking queue by instance. Meaning at most only one event is executed at the same time per instances. The pool size is >1. So if there is one instance keep connecting/disconnecting, wont impact others.